### DISCUSSION TITLE
Access in/out audio data

### DISCUSSION THREAD

**Discussion by franjf97** [pschatzmann/arduino-audio-tools#366] [Q&A] [answered]
Hi! I'm just starting with this Library setting up a few Stream examples. I managed to set up the i2s-filter-i2s Stream example as is shown here https://www.youtube.com/watch?v=D0EUDRo4Mx0&ab_channel=NA5Y.

Now I'm wondering if I could use the incoming audio data as a "raw variable" and try to implement filters or another process "manually", I don't know what type of data is a Stream in that case, but if I could use it as my input data for a certain function and pass that processed data to the audio sink it would be very helpfull for my project.

Thanks for everything!

**Comment by pschatzmann** [comment]
Did you read the Wiki ?  it is explained in the Introduction: https://github.com/pschatzmann/arduino-audio-tools/wiki

**Reply by franjf97** [reply]
I did. And looking at AudioCopy.h I think I have a brief idea of what copy() function does. I think I don't need to make my own implementation, but instead I'd like just to access the data that is going in and out that function. How can I do that?

**Reply by franjf97** [reply]
For example, something that gives me a callback when the in buffer transfer is complete and something similar with the out buffer would be a start, so I know when I can read that data.

**Comment by pschatzmann** [comment, marked as answer]
You are think way too complicated: In this document it is explained how the data format is and that you can handle it like e.g. a file. It also explains the data format: If you don't know how to do this you should probably not be using this framework.

Instead of using copy just do the following:
```
int16_t buffer[512];
int sample_size = sizeof(int16_t); // if bits_per_sample is set to 16
size_t eff_samples = in.readBytes((uint8_t*)buffer,512*sample_size) / sample_size;
// do with you data whatever you want for eff_samples
..
out.write((uint8_t*)buffer, eff_samples*sample_size);
```

And of cause there is a callback. Just look it up yourself in the class documentation!
